### ISSUE TITLE
When using SNMPv3, some devices may show polling issues

### ISSUE THREAD

**Issue by bernisys** [Cacti/cacti#5797]
**Describe the bug**
Among the device types we are monitoring, there are a few which have major problems during polling cycles. So far mainly Juniper and Cisco ASA devices are not polled properly. The devices were all adapted by the corresponding change-device.php cli script, but i doubt that the script has done anything funny. In fact, the credentials seem to be working fine, because when we go to the "device edit" page in the Cacti GUI, we see the SNMP information coming just fine from all devices. So i assume there is something within spine that causes this issue, which is reacting differently than the PHP based parts in the GUI.

I suspect that there is something odd happening on SNMP side, which is causing spine to drop the information, or maybe a packet is not coming back properly. But on the capture i made i can see packets being sent and arriving properly (not 100% sure because i can't decode the content due to the encryption),. What's significant is that the packets come always in pairs with identical data portion encapsulated. Is that normal behavior?

**To Reproduce**
1. Go to '...' ebay and buy an old Cisco ASA ... 
Ok, just kidding, not sure if this is really reproducible. The interesting part is that outside of spine everything looks fine.

**Server (please complete the following information):**
 - OS: RHEL
 - Version 7.9

**Compiling (please complete the following information):**
 - compiler: gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
 - autoconf: autoconf (GNU Autoconf) 2.69
 - glibc: 2.17-326.el7_9
 - source: release 1.2.17 / 23 / 25  (tested with all 3 versions)

**Additional context**
Did some spine runs in debug mode and what i saw is that they all have rather large delays between the "Updating Full System Information Table" and "RECACHE: Processing 3 items in the auto reindex cache" part - the time span of the gaps depends on the timeout value (tested with 1500ms and 2500ms). Here's a shortened version, maybe you already get some ideas and if you guide me a bit, i could add some more debugging code in specific places, recompile and gather some details.

```
      ... skipping the init sequence, which looks good
1719990805.378535 Total[0.1096] Device[303] Checking for System Information Update
1719990805.378570 Total[0.1096] Device[303] Updating Full System Information Table
      (15s spent here with no output - timeout is 2500ms, means 6x timeout reached somewhere?)
1719990820.842316 Total[0.1172] DEBUG: Device[303] HT[1] RECACHE: Processing 3 items in the auto reindex cache for 'xx.xx.xx.xx'
1719990820.842428 Total[0.1181] Device[303] HT[1] DQ[1] Legacy Uptime Result: 906303816, Is Numeric: 1
1719990820.842466 Total[0.1188] Device[303] HT[1] DQ[1] Extended Uptime Result: 20023448200, Is Numeric: 1
1719990820.842572 Total[0.1188] Device[303] HT[1] DQ[32] RECACHE OID: .1.3.6.1.2.1.1.3.0, (assert: 20023438800 < output: 20023448200)
1719990820.842608 Total[0.1199] Device[303] HT[1] NOTE: There are '31' Polling Items for this Device
1719990820.842645 Total[0.1199] The executable is '/usr/bin/perl' in 'perl /opt/SP/cacti/cacti/scripts/ping.pl 'xxx.xxx.xxx.xxx''
1719990820.842748 Total[0.1202] DEBUG: Device[303] DEBUG: The NIFTY POPEN returned the following File Descriptor 46
1719990820.842784 Total[0.1342] Device[303] HT[1] DS[4517] TT[14.26] SCRIPT: perl /opt/SP/cacti/cacti/scripts/ping.pl 'xxx.xxx.xxx.xxx', output: 1.87
1719990820.842819 Total[0.1342] Device[303] INFO: SNMP Device 'xxx.xxx.xxx.xxx:161' has a timeout of 2500000 (2500), with 3 retries
1719990820.842854 Total[0.5794] WARNING: Device[303] polling sleeping while waiting for 1 Threads to End
1719990820.842890 Total[1.0795] WARNING: Device[303] polling sleeping while waiting for 1 Threads to End
      ... lots of those thread-waits
1719990820.843559 Total[10.1447] Device[303] HT[1] DS[14613] WARNING: SNMP timeout detected [2500 ms], ignoring host 'xxx.xxx.xxx.xxx'
1719990820.843595 Total[10.1447] Device[303] HT[1] DS[14613] TT[10010.47] SNMP: v3: xxx.xxx.xxx.xxx, dsname: asa_connections, oid: 1.3.6.1.4.1.9.9.147.1.2.2.2.1.5.40.6, value: U
1719990820.843631 Total[10.5820] WARNING: Device[303] polling sleeping while waiting for 1 Threads to End
      ... lots of those thread-waits again, then another delay of 5s
**1719990820**.843959 Total[15.0834] WARNING: Device[303] polling sleeping while waiting for 1 Threads to End
**1719990825**.845514 Total[15.5835] WARNING: Device[303] polling sleeping while waiting for 1 Threads to End
      ... more thread-waits
1719990825.846107 Total[20.0848] WARNING: Device[303] polling sleeping while waiting for 1 Threads to End
1719990825.846145 Total[20.1546] Device[303] HT[1] DS[14614] WARNING: SNMP timeout detected [2500 ms], ignoring host 'xxx.xxx.xxx.xxx'
1719990825.846178 Total[20.1546] Device[303] HT[1] DS[14614] TT[20020.33] SNMP: v3: xxx.xxx.xxx.xxx, dsname: asa_active_sessions, oid: 1.3.6.1.4.1.9.9.392.1.3.1.0, value: U
1719990825.846212 Total[20.1548] DEBUG: Setting up writes to remote database
1719990825.846246 Total[20.1549] Device[303] HT[1] Total Time: 20 Seconds
1719990825.846280 Total[20.1555] DEBUG: Freeing Local Pool ID 0
1719990825.846315 Total[20.1555] DEBUG: Freeing Remote Pool ID 0
1719990825.846351 Total[20.1555] DEBUG: Device[303] HT[1] DEBUG: HOST COMPLETE: About to Exit Device Polling Thread Function
1719990825.846388 Total[20.5850] The final count of Threads is 0
1719990825.846423 Total[20.5850] INFO: Thread complete for Device[303] and 0 to 0 sources
1719990825.846459 Total[20.5850] DEBUG: Closing Local Connection Pool ID 0
```

**Comment by bernisys** [comment]
Found a post in the forum https://forums.cacti.net/viewtopic.php?t=62102
Larry suggests (and i comment):
- With SNMPv3, you have to be real careful.    (yeah, tell me about it ... >_<)
- make sure your Net-SNMP is v5.8++   (that's a hard one, as RHEL 7.9 is still shipping 5.7.2, but we can try if we find something in the SCL)
- make sure you are NOT running php-snmp   (we're using spine of course)
- make sure you are running Spine 1.2.21+   (we are)

But still, i find it quite interesting, that >3000 other devices work just fine out of the box, while Junipers and this one ASA we tried for testing seem to have problems. And i can't really see a problem with "two indexes" as indicated by the original poster.
We will try to create the ASA device from scratch, in some other cases this helped, but the Junipers could not be fixed with this last hope procedure.

In any case I will try to follow the hint and get hold of an alternative version of the snmp tools and try to compile it and test with the new one, but this needs some time.

**Comment by TheWitness** [comment]
Please keep us posted.  I'm busy on other things, but will double back to spine eventually.

**Comment by TheWitness** [comment]
I think we fixed the custom Cisco version of SNMPv3 in 1.2.26 or 1.2.27.  Check the changelog.

**Comment by TheWitness** [comment]
There are two issues blocking the adoption of 1.2.27 in my opinion. One of them is related to TCP ping and the other is related to boost template ordering on dynamic graph rendering. 

I have been spending a lot of time on the flowview plug-in and should be winding down in a week or two to refocus on t-hold and cacti/spine.

**Comment by bernisys** [comment]
Hi Larry, thanks for the heads-up.

From our side, we had some success for other devices, sometimes re-scanning the individual data-queries made things look a bit better, but for JunOS we still see the same issue on all devices. As if those don't want to communicate with cacti on v3 .. switching back to v2 immediately solves the problem. I wondered first if it might be necessary to provide an engine-ID, but then again, the system MIB works fine in the GUI, and with snmpwalk on the shell the data is rushing in immediately.

Will keep you posted if we find anything that helps!
Have a nice weekend!

**Comment by TheWitness** [comment]
@bernisys,

That sounds more like a device level ACL that's blocking access to the broader information. 

It's the same with Net-SNMP. In the default install you only get access to the system mibs and nothing else. 

Larry

**Comment by bernisys** [comment]
@TheWitness nah, we tested that, sorry i didn't note that above.
With snmpwalk on the shell and the same credentials as in Cacti, everything works just fine.
We were even down to packet level debugging already and see no real cause ...

**Comment by TheWitness** [comment]
@bernisys,

Provide the

1. Security Level
2. Privacy Protocol
3. Auth Protocol

Are they all the same?  Do other devices work with these same settings?  Are the broken devices maybe expecting Diffie-Helman key exchange?

**Comment by bernisys** [comment]
@TheWitness,
For the JunOS type we use:
    1. Security Level = authPriv
    2. Privacy Protocol = AES (aes128 on shell)
    3. Auth Protocol = SHA (sha1 on shell)

Fun fact. We have several thousand devices on sha/aes, and most of them are working just fine.
There are also several hundred on md5/des, which all work fine as far as i can see.

Could it have to do with the intermediate accidental removal of the legacy AES protocol? But the device should use AES128, and in the GUI it delivers the System-MIB data properly (though this comes via PHP poll, not spine). I will try with spine 1.2.27 and see if it works better with that version, but I need to compile and distribute first.

**Comment by TheWitness** [comment]
Keep us posted.  Maybe.  Was it working before?

**Comment by bernisys** [comment]
We were just forced to switch from V2 to V3 for the whole installation, so this is our first experience more or less with those specific devices. There are other devices (mainly Cisco switches) which we already had switched to V3 some time ago, they are working quite fine so far.

**Comment by bernisys** [comment]
I just collected some overview about our SNMP versions and the different templates we use (stripped out the irrelevant ones).
As you can see, there are several where the transition is not completed yet, mostly due to credentials deployment on the devices. but the major portion is running on V3 already, and mostly doing just fine as far as i could see.

```
select host_template_id as tpl_id, host_template.name as template_name, snmp_version, count(1) as count
from host join host_template on host_template_id = host_template.id group by host_template_id, snmp_version;
+--------+-----------------------------------------------+--------------+-------+
| tpl_id | template_name                                 | snmp_version | count |
+--------+-----------------------------------------------+--------------+-------+
|      1 | Cisco - Router                                |            2 |    21 |
|      1 | Cisco - Router                                |            3 |    21 |
|      2 | Generic SNMP Device                           |            2 |     1 |
|      3 | Linux Remote Server                           |            2 |     1 |
|      3 | Linux Remote Server                           |            3 |     4 |
|      6 | Cisco - Nexus Switch                          |            2 |    54 |
|      6 | Cisco - Nexus Switch                          |            3 |  1303 |
|      7 | Cisco - Switch - IOS 12.2                     |            2 |    72 |
|      7 | Cisco - Switch - IOS 12.2                     |            3 |   163 |
|      8 | HP - Switch 12k                               |            2 |     3 |
|      8 | HP - Switch 12k                               |            3 |     3 |
|      9 | Cisco - Catalyst 37xx switch                  |            2 |   130 |
|      9 | Cisco - Catalyst 37xx switch                  |            3 |    95 |
|     12 | Juniper - JunOS - VIP                         |            2 |    56 |
|     12 | Juniper - JunOS - VIP                         |            3 |    19 |
|     14 | Fortinet - Fortigate Proxy                    |            2 |    11 |
|     14 | Fortinet - Fortigate Proxy                    |            3 |    71 |
|     15 | Cisco - ASA                                   |            2 |    12 |
|     16 | Cisco - ASA 9.10                              |            2 |    38 |
|     16 | Cisco - ASA 9.10                              |            3 |     6 |
|     17 | F5 - BigIP Load Balancer                      |            2 |     4 |
|     17 | F5 - BigIP Load Balancer                      |            3 |   105 |
|     18 | Checkpoint (Nokia IPSO)                       |            2 |    30 |
|     19 | Juniper - Netscreen                           |            2 |     9 |
|     20 | VF-FW-Bluecoat Proxy                          |            2 |     6 |
|     21 | Juniper - SPACE                               |            2 |     6 |
|     22 | VF-SW-Gigamon                                 |            3 |    32 |
|     24 | Cisco - Nexus Switch - 5_6K                   |            2 |     8 |
|     24 | Cisco - Nexus Switch - 5_6K                   |            3 |   741 |
|     26 | Cisco ASR 9K                                  |            2 |     3 |
|     26 | Cisco ASR 9K                                  |            3 |    99 |
|     27 | VF-CTL-Cisco-APIC                             |            3 |    51 |
|     28 | VF-NTP-Meinberg                               |            2 |    12 |
|     28 | VF-NTP-Meinberg                               |            3 |     1 |
|     29 | VF-OIT-Cisco-WLC                              |            2 |     1 |
|     29 | VF-OIT-Cisco-WLC                              |            3 |    28 |
|     30 | Infloblox                                     |            3 |     6 |
|     31 | VF-PFS-Netscout                               |            2 |    24 |
|     32 | VF-OIT-Cisco Switch                           |            2 |     3 |
|     32 | VF-OIT-Cisco Switch                           |            3 |   386 |
|     33 | VF-OIT-Cisco Router                           |            3 |     9 |
|     34 | Cisco - LB - ACE (Main context)               |            2 |    22 |
|     34 | Cisco - LB - ACE (Main context)               |            3 |     1 |
|     35 | Cisco - LB - ACE (Virtual Context)            |            2 |   187 |
|     35 | Cisco - LB - ACE (Virtual Context)            |            3 |     1 |
|     36 | VF-OIT-Cisco - Switch - IOS 12.2              |            3 |     6 |
|     37 | Juniper - JunOS 3K - VIP                      |            2 |    17 |
|     37 | Juniper - JunOS 3K - VIP                      |            3 |     5 |
+--------+-----------------------------------------------+--------------+-------+
```

**Comment by TheWitness** [comment]
Could there be a time sync issue?  It's one weakness of V3.

**Comment by bernisys** [comment]
from what i can say. the servers and devices are all NTP controlled, so the timestamps should be all within the range of accuracy of the NTP protocol. I will check with the team, but i believe time discrepancies are not the issue. Good point anyway.

**Comment by bernisys** [comment]
@TheWitness
Firewall team finally has some response from Junpier!
They told us that on the firewall itself a unique engine ID needs to be configured.
This was now done for testing on two of the Juniper devices, which instantly started to deliver data when switched over to v3.
In Cacti however there's no need to enter the Engine ID, it seems to be auto-detected properly.

Also there is a certain process to do that on the device - otherwise it will fail!
* The engine ID needs to be changed/added first.
* And then you need to commit this change.
* Then all existing users need to be removed.
* commit.
* And only after that, you can add users.
* commit.

We will observe the behavior now for a certain amount of time, and if we consider it stable, we can close this ticket.
I think you can/should tag it with "device issue" or something similar.

**Comment by TheWitness** [comment]
Sounds like a game of twister, only with tech and not other teenagers...

**Comment by bernisys** [comment]
A bit yes ... but we're happy to have a solution finally - and also happy that it's not on Cacti side :)

**Comment by bernisys** [comment]
What still keeps me pondering about:
If the engine ID is the culprit, why are live graphs then working fine?
And why is the System MIB showing properly in the device edit page?
What's the actual difference between the polling and the remote retrieval of the data on demand?
Is it only that spine uses the compiled system libs while the on-demand stuff is probably done via php libraries?
Are those methods so much different from each other that one fails and the other does not?

**Comment by bernisys** [comment]
@TheWitness 
There seems to be more to it .. and now it gets even more interesting.

We now had a similar problem on a FortiGate when switching off v2 on the device while v3 polling was already activated and working fine.

Here's what i detected:
* some of the queries were still using SNMP v2 protocol
* i could find out that this was mainly related to the **down detection** method - tcpdump shows that it requests the system MIB in v2 mode step by step
* when changing an existing device in the GUI from v2 to v3, the old v2 community is still present in the DB and seems to be used in the down detection process
* I have deleted the v2 community from the DB on both master and remote poller - this did not fully help yet, something is still in the DB, which makes it use v2 but now with an empty community string
* going to the device edit page and hitting "save" after the removal from the DB seems to do finally the trick in the end
* still there are a few elements which are not delivered properly by the device (so far i can only see the NIC discards and error counters failing)

but doing this "manual save" on >3000 devices is an awful lot of overhead ...


I took a tcpdump. Significant notes:
* First thing that seems odd is the "C="-On" in the first lines where the system MIB is requested
* Then a few v3 polls are running and working fine apparently
* In the middle when it tries to poll the errors/discards count from the interfaces MIB2, it goes back to v2 again

```
13:55:33.513862 IP poller_ip.41363 > device_ip.161:  C="-On" GetNextRequest(26)  .1.3.6.1.2.1.1
13:55:34.514944 IP poller_ip.41363 > device_ip.161:  C="-On" GetNextRequest(26)  .1.3.6.1.2.1.1
13:55:35.516025 IP poller_ip.41363 > device_ip.161:  C="-On" GetNextRequest(26)  .1.3.6.1.2.1.1
13:55:36.517104 IP poller_ip.41363 > device_ip.161:  C="-On" GetNextRequest(26)  .1.3.6.1.2.1.1
13:55:37.518270 IP poller_ip.41363 > device_ip.161:  C="-On" GetNextRequest(26)  .1.3.6.1.2.1.1
13:55:38.519327 IP poller_ip.41363 > device_ip.161:  C="-On" GetNextRequest(26)  .1.3.6.1.2.1.1
13:55:39.716460 IP poller_ip.37981 > device_ip.161:  F=r U="" E= C="" GetRequest(14)
13:55:39.845444 IP device_ip.161 > poller_ip.37981:  F= U="" E=<removed> C="" Report(32)  .1.3.6.1.6.3.15.1.1.4.0=27861
13:55:39.845651 IP poller_ip.37981 > device_ip.161:  F=apr U="testuser" [!scoped PDU] ....
13:55:39.845972 IP device_ip.161 > poller_ip.37981:  F=ap U="testuser" [!scoped PDU] ....
13:55:39.846453 IP poller_ip.37981 > device_ip.161:  F=apr U="testuser" [!scoped PDU] ....
13:55:39.846728 IP device_ip.161 > poller_ip.37981:  F=ap U="testuser" [!scoped PDU] ....
13:55:39.846806 IP poller_ip.37981 > device_ip.161:  F=apr U="testuser" [!scoped PDU] ....
13:55:39.847054 IP device_ip.161 > poller_ip.37981:  F=ap U="testuser" [!scoped PDU] ....
13:55:39.883251 IP poller_ip.34712 > device_ip.161:  F=r U="" E= C="" GetRequest(14)
13:55:39.883717 IP device_ip.161 > poller_ip.34712:  F= U="" E=<removed> C="" Report(32)  .1.3.6.1.6.3.15.1.1.4.0=27862
13:55:39.883803 IP poller_ip.34712 > device_ip.161:  F=apr U="testuser" [!scoped PDU] ....
13:55:39.884102 IP device_ip.161 > poller_ip.34712:  F=ap U="testuser" [!scoped PDU] ....
13:55:39.884294 IP poller_ip.51748 > device_ip.161:  C="v2community" GetRequest(408)  .1.3.6.1.2.1.2.2.1.13.24 .1.3.6.1.2.1.2.2.1.19.24 .1.3.6.1.2.1.2.2.1.14.24 .1.3.6.1.2.1.2.2.1.20.24 .1.3.6.1.2.1.2.2.1.13.25 .1.3.6.1.2.1.2.2.1.19.25 .1.3.6.1.2.1.2.2.1.14.25 .1.3.6.1.2.1.2.2.1.20.25 .1.3.6.1.2.1.2.2.1.13.26 .1.3.6.1.2.1.2.2.1.19.26 .1.3.6.1.2.1.2.2.1.14.26 .1.3.6.1.2.1.2.2.1.20.26 .1.3.6.1.2.1.2.2.1.13.33 .1.3.6.1.2.1.2.2.1.19.33 .1.3.6.1.2.1.2.2.1.14.33 .1.3.6.1.2.1.2.2.1.20.33 .1.3.6.1.2.1.31.1.1.1.6.38 .1.3.6.1.2.1.31.1.1.1.10.38 .1.3.6.1.2.1.31.1.1.1.6.40 .1.3.6.1.2.1.31.1.1.1.10.40 .1.3.6.1.2.1.31.1.1.1.6.42 .1.3.6.1.2.1.31.1.1.1.10.42 .1.3.6.1.2.1.31.1.1.1.6.44 .1.3.6.1.2.1.31.1.1.1.10.44
13:55:40.384863 IP poller_ip.51748 > device_ip.161:  C="v2community" GetRequest(408)  .1.3.6.1.2.1.2.2.1.13.24 .1.3.6.1.2.1.2.2.1.19.24 .1.3.6.1.2.1.2.2.1.14.24 .1.3.6.1.2.1.2.2.1.20.24 .1.3.6.1.2.1.2.2.1.13.25 .1.3.6.1.2.1.2.2.1.19.25 .1.3.6.1.2.1.2.2.1.14.25 .1.3.6.1.2.1.2.2.1.20.25 .1.3.6.1.2.1.2.2.1.13.26 .1.3.6.1.2.1.2.2.1.19.26 .1.3.6.1.2.1.2.2.1.14.26 .1.3.6.1.2.1.2.2.1.20.26 .1.3.6.1.2.1.2.2.1.13.33 .1.3.6.1.2.1.2.2.1.19.33 .1.3.6.1.2.1.2.2.1.14.33 .1.3.6.1.2.1.2.2.1.20.33 .1.3.6.1.2.1.31.1.1.1.6.38 .1.3.6.1.2.1.31.1.1.1.10.38 .1.3.6.1.2.1.31.1.1.1.6.40 .1.3.6.1.2.1.31.1.1.1.10.40 .1.3.6.1.2.1.31.1.1.1.6.42 .1.3.6.1.2.1.31.1.1.1.10.42 .1.3.6.1.2.1.31.1.1.1.6.44 .1.3.6.1.2.1.31.1.1.1.10.44
13:55:40.885435 IP poller_ip.51748 > device_ip.161:  C="v2community" GetRequest(408)  .1.3.6.1.2.1.2.2.1.13.24 .1.3.6.1.2.1.2.2.1.19.24 .1.3.6.1.2.1.2.2.1.14.24 .1.3.6.1.2.1.2.2.1.20.24 .1.3.6.1.2.1.2.2.1.13.25 .1.3.6.1.2.1.2.2.1.19.25 .1.3.6.1.2.1.2.2.1.14.25 .1.3.6.1.2.1.2.2.1.20.25 .1.3.6.1.2.1.2.2.1.13.26 .1.3.6.1.2.1.2.2.1.19.26 .1.3.6.1.2.1.2.2.1.14.26 .1.3.6.1.2.1.2.2.1.20.26 .1.3.6.1.2.1.2.2.1.13.33 .1.3.6.1.2.1.2.2.1.19.33 .1.3.6.1.2.1.2.2.1.14.33 .1.3.6.1.2.1.2.2.1.20.33 .1.3.6.1.2.1.31.1.1.1.6.38 .1.3.6.1.2.1.31.1.1.1.10.38 .1.3.6.1.2.1.31.1.1.1.6.40 .1.3.6.1.2.1.31.1.1.1.10.40 .1.3.6.1.2.1.31.1.1.1.6.42 .1.3.6.1.2.1.31.1.1.1.10.42 .1.3.6.1.2.1.31.1.1.1.6.44 .1.3.6.1.2.1.31.1.1.1.10.44
13:55:41.386006 IP poller_ip.51748 > device_ip.161:  C="v2community" GetRequest(408)  .1.3.6.1.2.1.2.2.1.13.24 .1.3.6.1.2.1.2.2.1.19.24 .1.3.6.1.2.1.2.2.1.14.24 .1.3.6.1.2.1.2.2.1.20.24 .1.3.6.1.2.1.2.2.1.13.25 .1.3.6.1.2.1.2.2.1.19.25 .1.3.6.1.2.1.2.2.1.14.25 .1.3.6.1.2.1.2.2.1.20.25 .1.3.6.1.2.1.2.2.1.13.26 .1.3.6.1.2.1.2.2.1.19.26 .1.3.6.1.2.1.2.2.1.14.26 .1.3.6.1.2.1.2.2.1.20.26 .1.3.6.1.2.1.2.2.1.13.33 .1.3.6.1.2.1.2.2.1.19.33 .1.3.6.1.2.1.2.2.1.14.33 .1.3.6.1.2.1.2.2.1.20.33 .1.3.6.1.2.1.31.1.1.1.6.38 .1.3.6.1.2.1.31.1.1.1.10.38 .1.3.6.1.2.1.31.1.1.1.6.40 .1.3.6.1.2.1.31.1.1.1.10.40 .1.3.6.1.2.1.31.1.1.1.6.42 .1.3.6.1.2.1.31.1.1.1.10.42 .1.3.6.1.2.1.31.1.1.1.6.44 .1.3.6.1.2.1.31.1.1.1.10.44
13:55:41.908217 IP poller_ip.52686 > device_ip.161:  F=r U="" E= C="" GetRequest(14)
13:55:41.908617 IP device_ip.161 > poller_ip.52686:  F= U="" E=<removed> C="" Report(32)  .1.3.6.1.6.3.15.1.1.4.0=27863
13:55:41.908715 IP poller_ip.52686 > device_ip.161:  F=apr U="testuser" [!scoped PDU] ....
13:55:41.908993 IP device_ip.161 > poller_ip.52686:  F=ap U="testuser" [!scoped PDU] ....
```

**Comment by bernisys** [comment]
@TheWitness 

There are several entries in poller_item with SNMP version 2, which won't go away even if i run the poller cache rebuild or re-indexing for the device. Those are probably the ones causing the problems. But how can i get rid of those?

Can i find out somehow, if those are orphans?

I know we had error/discard graphs once activated for the specific device and they had worked. At some point in time those seem to have vanished. Feels like cacti lost the link between the graph, the device and the data source. But which kind of operation could have caused that? I had this feeling already for a long time, that something is not completely right, but i couldn't really point to it.

Shall i open a related ticket in the Cacti tracker?
As this seems to have nothing to do with spine any more, but it starts looking like a cacti related thing.

**Comment by TheWitness** [comment]
That's quite odd.  They must be old orphaned entries from when Cacti allowed the Graphs to be deleted, and allowed the unused Data Sources to remain.

Try running the repair database CLI.  There is a special flag to do the more esoteric cleanups.  Post the dry run results.

**Comment by bernisys** [comment]
Usually we don't keep datasources, the preset is to have the delete DS switched on all the time, so it should (in theory...) not have happened that graphs/datasources were kept when something got deleted. But well, you never know.

Here's the result of the script run:
```
Repairing All Cacti Database Tables

NOTE: Running some Data Query repair scripts
NOTE: Checking for Invalid Cacti Templates
NOTE: 1 Invalid Data Input Rows found in Data Templates
NOTE: 216 Invalid Data Input Data Rows based upon template mappings found in Data Templates

WARNING: Cacti Template Problems found in your Database.  Using the '--force' option will remove
the invalid records.  However, these changes can be catastrophic to existing data sources.  Therefore, you
should contact your support organization prior to proceeding with that repair.
```

I think for the template cleanups, i should check first, which templates would be affected and definitely take a DB backup...
One of our users had noticed that one template for interfaces was borked some time ago, there was a duplication of all lines in the legend and all graph items for all devices in all diagrams with this template. I had already checked the tables and saw that the actual graph of one specific device was containing template IDs of zero everywhere, so the contents of that graph were taken accidentally as additional template lines. And on propagation, all interface graphs were adding those dupes.

Maybe there is a way to check that during graph creation - but that's another story. I can open a ticket if you like and describe what we detected, but so far i did not see that happening again.

**Comment by TheWitness** [comment]
Yeah I guess that's fine the other thing is I'm just wondering if you have a few data input methods that are broken and therefore the settings are not being populated from The Host. 

So it's a tricky query to find these things and generally can happen if you have say the in the data template the SNMP version checked with a check mark which means you can override the devices setting so there's some queries that we could run to expose the problem and I guess there's a CLI utility that we could develop to fix that. 

This goes back to a debate I had with Ian some years ago where he wanted every specific data source to be able to have its own SNMP information because at that time there were a lot of java-based SNMP agents that ran on separate ports with separate community strings as opposed to extending the default SNMP agent. 

What it means in the end there's a lot of data bloat in the database and that's in the table data input data which you have 216 defective rows but I expect that there is probably something else going on in there especially if this is an older install.

**Comment by bernisys** [comment]
Ok i see .. so let's put a check mark at "[x] it's complicated" ...

I started seeing the complexity already when i was digging into the DB and perts the code, but i guess that's just the tip of the iceberg which i saw so far. The DB repair queries are giving already some insight into what's generally going wrong, but that's also only a part of the whole picture that i am slowly getting.

i will try and find out where those 216 are coming from, and also i need to check, if there are any orphans in the poller_item table like those still entries pointing to snmpv2 while the device is on v3 and those graphs are somehow gone ...
any hints for a neat query?
i would - from my current PoV - just relate the host-ID from "host" with the host-ID fro m"poller_item" and check if the snmp versions are different between the host and the items and see what's showing up + try cleaning them up.
Can i somehow find out what entries would have been generated in poller_item just by looking into the graph table? Somewhere should be a connection between both - i always forget the relations, they are partially quite complex...

hints welcome, then i can try around here and give pointers which might help creating a script that can help others to find and clean up stale data.

**Comment by TheWitness** [comment]
The simplest one is this....

```sql
SELECT host_id, dl.data_template_id, COUNT(*) AS busted
FROM poller_item AS pi
INNER JOIN data_local AS dl
ON dl.id = pi.local_data_id
LEFT JOIN host AS h
ON h.id = pi.host_id
WHERE h.snmp_version != pi.snmp_version
GROUP BY h.id, dl.data_template_id
```

That'll show the byorked Data Templates.  You can extend that query to find the Graphs impacted too.

**Comment by TheWitness** [comment]
From there you can:

1. Edit the Data Template to see what's up.
2. Determine the scope of the issue.
3. Work out the details of next steps.  Maybe a repair utility.

Make a dump of the database before making any changes as that'll be the test schema for the repair CLI development.

**Comment by bernisys** [comment]
okay, 597 rows in set (2.429 sec)
that's quite a lot to analyze, and there are quite a lot of entries listed in the "busted" column .. phew ...
Thanks so far for that quick hint with the query, i will see how i can proceed and check the results in more depth.

```
+---------+------------------+--------+
| host_id | data_template_id | busted |
+---------+------------------+--------+
|     118 |                2 |     76 |
|     128 |                2 |    102 |
|     140 |                2 |     62 |
|     156 |                2 |    112 |
|     161 |                2 |    108 |
|     162 |                2 |    118 |
|     165 |                2 |     80 |
|     166 |                2 |     80 |
  ... lots more
```

**Comment by TheWitness** [comment]
What is Data Template ID 2?

**Comment by bernisys** [comment]
Here's a complete overview without the host IDs .. it's quite a lot. Somewhere around 28-29k items in total.

```
SELECT dl.data_template_id, dt.name, COUNT(*) AS busted FROM poller_item AS pi
  INNER JOIN data_local AS dl ON dl.id = pi.local_data_id
  LEFT JOIN host AS h ON h.id = pi.host_id
  INNER JOIN data_template AS dt ON dt.id=dl.data_template_id
  WHERE h.snmp_version != pi.snmp_version GROUP BY dl.data_template_id;
+------------------+-----------------------------------------------------------------------+--------+
| data_template_id | name                                                                  | busted |
+------------------+-----------------------------------------------------------------------+--------+
|                2 | Interface - Traffic                                                   |  20640 |
|                3 | Interface - Errors/Discards                                           |    356 |
|                6 | Unix - Ping Host                                                      |     24 |
|               32 | Host MIB - CPU Utilization                                            |      5 |
|               59 | Juniper - JunOS - SPU Usage Indexed - CPU / Memory                    |      4 |
|               64 | Juniper - JunOS - NAT-Pool Utilisation                                |      4 |
|               65 | Juniper - JunOS - OE CPU / Memory                                     |     16 |
|               72 | Cisco - ASA - CPU Load (before ASA v9.2)                              |     12 |
|               73 | Cisco - ASA - Connections Firewall (to be reviewed)                   |      4 |
|               74 | Cisco - ASA - Sessions - VPN Active                                   |      4 |
|               76 | Cisco - ASA - Connections Active                                      |      4 |
|              108 | F5 BigIP - Interface Errors/Drops/Collisions                          |    280 |
|              110 | F5 BigIP - TMM Memory Query                                           |     88 |
|              111 | F5 BigIP - TMM CPU                                                    |    132 |
|              146 | Cisco - Interface TxPause Frames                                      |   5222 |
|              169 | Cisco - IOS ASA ASR - Multi CPU Usage - 5min (new Cisco reviewed OID) |    311 |
|              170 | Cisco - IOS ASA ASR - Multi CPU Usage - 5min deprecated               |      7 |
|              189 | F5 BigIP - Interface Traffic                                          |    136 |
|              195 | Cisco - Generic - Memory Used                                         |     36 |
|              198 | Cisco LB - ACE - Context Bandwidth                                    |    102 |
|              199 | Cisco LB - ACE - Context Connections Concurrent                       |    102 |
|              200 | Cisco LB - ACE - Context Connections Proxied                          |    102 |
|              201 | Cisco LB - ACE - Context Management                                   |    102 |
|              202 | Cisco LB - ACE - Context Sticky Entries                               |    102 |
|              212 | Cisco LB - ACE - Context Connection-Rate                              |    102 |
|              222 | Juniper - JunOS - Session-Creations (IPv4, IPv6, Total)               |      8 |
|              223 | Juniper - JunOS - Sessions Total (IPv4, IPv6, Total)                  |      8 |
|              224 | Juniper - JunOS - SPU Usage Indexed - Session Stats IPv4 IPv6         |      4 |
|              225 | Cisco - ASA - Sessions - VPN-SSL                                      |      4 |
|              226 | Cisco - ASA - Sessions - VPN-Web                                      |      4 |
|              240 | Device Polling Time                                                   |     21 |
+------------------+-----------------------------------------------------------------------+--------+
31 rows in set (2.332 sec)
```

**Comment by TheWitness** [comment]
We will need to go the dump route. I'll be in touch.

**Comment by TheWitness** [comment]
@bernisys, going to transfer this to Cacti.

**Comment by bernisys** [comment]
Follow-up:
  * The bug was mainly but not only related to some issues that were fixed between 1.2.18 and 1.2.27
  * Further adaptations were made in the code, finalizing those fixes (will be contained in 1.2.28)
  * A few steps were added in the "cli/repair_database.php" script (also contained in 1.2.28)
  * We will update Cacti soonest possible and then apply the DB repairs by running the new script
  * Prerequisites: Pull new versions of lib/functions.php, lib/utility.php and cli/repair_database.php
  * I will report here with the results of the upgrade
  
@TheWitness: As 1.2.28 is not tagged yet, please provide the commit numbers that contain the fixes in the files mentioned above.

**Comment by TheWitness** [comment]
You are correct.  If on 1.2.27, there are just a few changes required from the following commits.

This first one was to provide more robust repopulate performance.  It was 4x, but then I added detection of change.  So, not quite 4x now, but still fast.

https://github.com/Cacti/cacti/commit/03402112b80058fee1d9c9a01a4e45510b3bc3e6

This second creates a new function in lib/functions.php and improves the cli/repair_database.php specifically addressing your issues:

https://github.com/Cacti/cacti/commit/a510fbb14c4de8f92d2f3fbcb05ca91524d857c0

This next one fixes one typo...

https://github.com/Cacti/cacti/commit/3dc4b0e6cc3a7b5e779f82a59a0db03b6344b157

This one, though not required leverages the new function so as to not to change plugin tables.  Right now, I know of installations that leverage Latin1 for a reason as well as use MyISAM and Aria cause it's simply much faster.  So, don't try to convince users to migrate those table away.  Deal with the base tables only.  I suspect we will improve that in time too to be plugin specific.

https://github.com/Cacti/cacti/commit/774172cb80106fa8f90da3b1f59ecfd2b7fbffd9

Some typos and consistency in logging here:

https://github.com/Cacti/cacti/commit/cd40250237b390a7b480a4b5a74ec966f8f209ec

Finally correct some small logic bugs and properly re-update the orphan status if the index is found again during the repair.

https://github.com/Cacti/cacti/commit/d3bab5aa905824b49ff35c1e7cf4c2cd1a990dfb

So, basically you are correct.  Take the full files from the 1.2.x branch.  The one exception might be lib/functions.php.  You should do a diff first to see what else may have changed in that case.  

In the 1.2.28 release, it's mostly CVE's that rate sort of high, but require a nefarious admin to drive, and most Admin's have more creative ways to siphon things off.  I mean after all, this is open source.

**Comment by bernisys** [comment]
Update:
We just ran the repair script on our test system now, it says in the final line:
WARNING: Found 148495 and repaired 341680 Cacti database issues.

it sounded promising so far, but in the GUI **we don't see any hostnames and descriptions any more.**
Checking the DB, these entries are gone for good, all empty strings in there.
So I restored the backup we took, waiting for any ideas from your side what could be the issue.

meanwhile, i got some ideas to improve the script as well:
* it should write a log without having the user to use ">" or "tee" on the shell
* it should somehow write a summary about which graphs were attached to which device before the repair and add another report after the repair, so that we can see if any graphs were removed, that should have been kept
* in the final part where all the push-outs are happening, listing also the ID would be nice, as you can simply modify the device edit and graph view URLs if you know the ID, that improves the handling in the browser
